graveler/listing_iterator.go

```diff
@@ -0,0 +1,112 @@
+package graveler
+
+import (
+	"bytes"
+	"math"
+)
+
+// listingIterator implements a ListingIterator using a ValueIterator
+// assumes all values in valueIterator start with prefix
+type listingIterator struct {
+	valueIterator ValueIterator
+	delimiter     Key
+	prefix        Key
+	current       *Listing
+	nextFunc      func() bool
+}
+
+// getFollowingValue returns the following value (i.e will increase the last byte by 1)
+// in the following cases will return received value: empty value, the last byte is math.MaxUint8
+func getFollowingValue(value []byte) []byte {
+	if len(value) == 0 || value[len(value)-1] == math.MaxUint8 {
+		return value
+	}
+	copiedDelimiter := make([]byte, len(value))
+	copy(copiedDelimiter, value)
+	copiedDelimiter[len(copiedDelimiter)-1]++
+	return copiedDelimiter
+}
+
+func NewListingIterator(iterator ValueIterator, delimiter, prefix Key) ListingIterator {
+	l := &listingIterator{
+		valueIterator: iterator,
+		delimiter:     delimiter,
+		prefix:        prefix,
+	}
+	if len(delimiter) == 0 {
+		l.nextFunc = l.nextNoDelimiter
+	} else {
+		l.nextFunc = l.nextWithDelimiter
+	}
+	return l
+}
+
+func (l *listingIterator) nextNoDelimiter() bool {
+	hasNext := l.valueIterator.Next()
+	if !hasNext {
+		l.current = nil
+		return false
+	}
+	val := l.valueIterator.Value()
+	l.current = &Listing{
+		CommonPrefix: false,
+		Key:          val.Key,
+		Value:        val.Value,
+	}
+	return true
+}
+
+func (l *listingIterator) nextWithDelimiter() bool {
+	if l.current != nil && l.current.CommonPrefix {
+		nextKey := getFollowingValue(l.current.Key)
+		l.valueIterator.SeekGE(nextKey)
+	}
+	hasNext := l.valueIterator.Next()
+	if hasNext {
+		nextValue := l.valueIterator.Value()
+		l.current = l.getListingFromValue(nextValue.Value, nextValue.Key)
```

Comment:
```suggestion
		l.current = l.getListingFromValue(nextValue)
```
it can get ValueRecord as it is key+value